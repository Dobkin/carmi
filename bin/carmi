#!/usr/bin/env node
'use strict';
const commandLineArgs = require('command-line-args');
const carmi = require('../index');
const path = require('path');
const fs = require('fs-extra');
const {isUpToDate} = require('../src/analyze-dependencies');

const optionDefinitions = [
  { name: 'source', type: String, defaultOption: true, description: 'source filename, which exports a carmi model' },
  { name: 'output', type: String, defaultValue: '', description: 'output filename' },
  { name: 'cwd', type: String, defaultValue: '.', description: 'root directory for carmi files. Needed for babel config' },
  {
    name: 'compiler',
    type: String,
    defaultValue: 'optimizing',
    description: 'compiler version to use naive/simple/optimizing'
  },
  { name: 'debug', type: Boolean, defaultValue: false, description: 'add debug functions to the generated code' },
  { name: 'format', type: String, defaultValue: 'iife', description: 'output format - iife,cjs...' },
  { name: 'name', type: String, defaultValue: 'model', description: 'name of the output module/function' },
  { name: 'prettier', type: Boolean, defaultValue: false, description: 'run prettier on the output' },
  {
    name: 'cache',
    type: String,
    defaultValue: '',
    description: 'caches the generated code in the specified folder if the model did not change'
  },
  { name: 'no-cache', type: Boolean, defaultValue: false, description: 'ignore cache' },
  { name: 'stats', type: String, defaultValue: '', description: 'generate stats file' },
  { name: 'help', type: Boolean, defaultValue: false, description: 'shows this very help message and quits' },
  { name: 'ast', type: Boolean, defaultValue: false }
];

const HELP = `carmi [<options>] <src>
${optionDefinitions
  .filter(v => v.description)
  .map(opt => `--${opt.name} - ${opt.description}`)
  .join('\n')}
`;

async function run() {
  const options = commandLineArgs(optionDefinitions);
  if (options.help) {
    return console.log(HELP);
  }

  const cwd = options['cwd'] || process.cwd()
  require('@babel/register')({
    rootMode: 'upward', extensions: ['.ts', '.js'], cwd
  })
  
  const absPath = path.resolve(process.cwd(), options.source);

  const outFile = options.output || tempFile(absPath)
  // console.log(outFile)
  if (!options['no-cache'] && isUpToDate(absPath, outFile, options.stats)) {
    if (options.output) {
      console.log('output is up to date skipping carmi');
    } else {
      const out = await fs.readFile(outFile, 'utf-8')
      console.log(out);
    }
    return;
  }

  let model;
  try {
    model = require(absPath);
  } catch (e) {
    console.error(`failed to require ${options.source} ${e.stack}`);
    throw e;
  }
  const code = carmi.compile(model, options);
  if (!options.output) {
    console.log(code);
  }
  await fs.outputFile(outFile, code);
}

function tempFile(absPath) {
  const crypto = require('crypto');
  const hash = crypto.createHash('md5').update(absPath).digest("hex");
  // console.log(hash)
  return path.resolve('node_modules/.cache/carmi', hash)
}

run().catch(e => {
  console.log('error ' + e)
  process.exit(1)
});
