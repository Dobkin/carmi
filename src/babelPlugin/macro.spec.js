const pluginTester = require('babel-plugin-tester');
const plugin = require('babel-plugin-macros');
const path = require('path');
const babel = require('babel-core');

debugger
pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: path.resolve(__dirname, 'temp.js'),
    presets: ['react']
  },
  tests: {
    'carmi-react': `
    // @carmi
    import carmi from './macro'
    const { root } = require('../../index');
    const {createElement} = require('../../jsx');
    const todosList = <div>{root.get(0)}</div>;
    module.exports = {todosList};
`
  }
});

describe('Macro', () => {
  it('works', done => {
    const code = `
      const carmi = require('./macro')

      const modelBuilder = carmi\`
        const {root} = require('../..')
        module.exports = {all: root.get('list'), first: root.get('list').get(0)}
      \`

      const model = modelBuilder({ list: [1,2,3] })
      global.onModel(model)
    `;

    const transformedCode = babel.transform(code, {
      filename: __filename,
      plugins: [plugin]
    });

    global.onModel = model => {
      expect(model.all).toEqual([1, 2, 3]);
      expect(model.first).toBe(1);
      done();
    };

    // eval the script
    eval(transformedCode.code);
  });
});
