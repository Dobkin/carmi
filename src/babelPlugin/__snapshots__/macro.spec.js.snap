// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros carmi-react: carmi-react 1`] = `
"
// @carmi
import carmi from './macro'
const { root } = require('../../index');
const {createElement} = require('../../jsx');
const todosList = <div>{root.get(0)}</div>;
module.exports = {todosList};

      ↓ ↓ ↓ ↓ ↓ ↓

module.exports = function () {
  'use strict';
  function model($model, $funcLib, $batchingStrategy) {
    const $res = { $model };
    const $listeners = new Set();const $trackingMap = new WeakMap();const $trackedMap = new WeakMap();const $trackingWildcards = new WeakMap();const $invalidatedMap = new WeakMap();const $invalidatedRoots = new Set();$invalidatedRoots.$subKeys = {};$invalidatedRoots.$parentKey = null;$invalidatedRoots.$parent = null;let $first = true;let $tainted = new WeakSet();$invalidatedMap.set($res, $invalidatedRoots);const $uniquePersistentObjects = new Map();const getUniquePersistenObject = id => {
      if (!$uniquePersistentObjects.has(id)) {
        $uniquePersistentObjects.set(id, {});
      }return $uniquePersistentObjects.get(id);
    };const untrack = ($targetKeySet, $targetKey) => {
      const $tracked = $trackedMap.get($targetKeySet);if (!$tracked || !$tracked[$targetKey]) {
        return;
      }$tracked[$targetKey].forEach(({ $sourceObj, $sourceKey, $target }) => {
        const $trackingSource = $trackingMap.get($sourceObj);$trackingSource[$sourceKey].delete($target);
      });delete $tracked[$targetKey];
    };const invalidate = ($targetKeySet, $targetKey) => {
      if ($targetKeySet.has($targetKey)) {
        return;
      }$targetKeySet.add($targetKey);untrack($targetKeySet, $targetKey);if ($targetKeySet.$parent) {
        invalidate($targetKeySet.$parent, $targetKeySet.$parentKey);
      }
    };function setOnObject($target, $key, $val, $invalidates) {
      let $changed = false;let $hard = false;if ($invalidates && !$first) {
        if (typeof $target[$key] === 'object' && $target[$key] && $target[$key] !== $val) {
          $hard = true;
        }if ($hard || $target[$key] !== $val || typeof $val === 'object' && $tainted.has($val) || !$target.hasOwnProperty($key) && $target[$key] === undefined) {
          $changed = true;triggerInvalidations($target, $key, $hard);
        }
      }$target[$key] = $val;return $changed;
    }function setOnArray($target, $key, $val, $invalidates) {
      let $hard = false;if ($invalidates && !$first) {
        if (typeof $target[$key] === 'object' && $target[$key] && $target[$key] !== $val) {
          $hard = true;
        }if ($hard || $target[$key] !== $val || typeof $target[$key] === 'object' && $tainted.has($val) || !$target.hasOwnProperty($key) && $target[$key] === undefined) {
          triggerInvalidations($target, $key, $hard);
        }
      }$target[$key] = $val;
    }function triggerInvalidations($sourceObj, $sourceKey, $hard) {
      $tainted.add($sourceObj);const $track = $trackingMap.get($sourceObj);if ($track && $track.hasOwnProperty($sourceKey)) {
        $track[$sourceKey].forEach(($soft, $target) => {
          if (!$soft || $hard) {
            invalidate($target[0], $target[1]);
          }
        });
      }if ($trackingWildcards.has($sourceObj)) {
        $trackingWildcards.get($sourceObj).forEach($targetInvalidatedKeys => {
          invalidate($targetInvalidatedKeys, $sourceKey);
        });
      }
    }function getEmptyArray($invalidatedKeys, $targetKey, token) {
      const subKeys = $invalidatedKeys.$subKeys;const $cachePerTargetKey = subKeys[$targetKey] = subKeys[$targetKey] || new Map();if (!$cachePerTargetKey.has(token)) {
        $cachePerTargetKey.set(token, []);
      }return $cachePerTargetKey.get(token);
    }function array($invalidatedKeys, key, newVal, identifier, len, invalidates) {
      const res = getEmptyArray($invalidatedKeys, key, identifier);invalidates = invalidates && res.length === len;for (let i = 0; i < len; i++) {
        setOnArray(res, i, newVal[i], invalidates);
      }return res;
    }const $callCache = new WeakMap();function call($invalidatedKeys, key, newVal, identifier, len, invalidates) {
      const arr = getEmptyArray($invalidatedKeys, key, identifier);invalidates = true;for (let i = 0; i < len; i++) {
        setOnArray(arr, i, newVal[i], invalidates);
      }if (!$callCache.has(arr) || $tainted.has(arr)) {
        $callCache.set(arr, $funcLib[arr[0]].apply($res, arr.slice(1)));
      }return $callCache.get(arr);
    }$invalidatedRoots.add('todosList');function $todosListBuild() {
      const key = 'todosList';const $invalidatedKeys = $invalidatedRoots;const newValue = call($invalidatedKeys, key, [\\"createElement\\", $res[\\"$array_temp_4_19_1\\"]], getUniquePersistenObject(4), 2, false);setOnObject($res, 'todosList', newValue, false);$invalidatedRoots.delete('todosList');return $res.todosList;
    }const array$2Token = getUniquePersistenObject(2);const array$2Args = 3;$invalidatedRoots.add('$array_temp_4_19_1');function $array_temp_4_19_1Build() {
      const key = '$array_temp_4_19_1';const $invalidatedKeys = $invalidatedRoots;const newValue = array($invalidatedKeys, key, [\\"div\\", null, $model[0]], array$2Token, array$2Args, false);setOnObject($res, '$array_temp_4_19_1', newValue, false);$invalidatedRoots.delete('$array_temp_4_19_1');return $res.$array_temp_4_19_1;
    }let $inBatch = false;let $batchPending = [];function recalculate() {
      if ($inBatch) {
        return;
      }$invalidatedRoots.has('$array_temp_4_19_1') && $array_temp_4_19_1Build();$invalidatedRoots.has('todosList') && $todosListBuild();$tainted = new WeakSet();$first = false;$listeners.forEach(callback => callback());if ($batchPending.length) {
        $res.$endBatch();
      }
    }Object.assign($res, {}, { $startBatch: () => $inBatch = true, $endBatch: () => {
        $inBatch = false;if ($batchPending.length) {
          $batchPending.forEach(({ func, args }) => {
            func.apply($res, args);
          });$batchPending = [];recalculate();
        }
      }, $runInBatch: func => {
        $res.$startBatch();func();$res.$endBatch();
      }, $addListener: func => {
        $listeners.add(func);
      }, $removeListener: func => {
        $listeners.delete(func);
      } });recalculate();return $res;
  }return model;
}();
"
`;

exports[`macros magic-comment: magic-comment 1`] = `
"
// @carmi
import carmi from './macro'
const { root } = require('../../index');
module.exports = { first: root.get(0), second: root.get(1) };

      ↓ ↓ ↓ ↓ ↓ ↓

module.exports = function () {
  'use strict';
  function model($model, $funcLib, $batchingStrategy) {
    const $res = { $model };const $listeners = new Set();const $trackingMap = new WeakMap();const $trackedMap = new WeakMap();const $trackingWildcards = new WeakMap();const $invalidatedMap = new WeakMap();const $invalidatedRoots = new Set();$invalidatedRoots.$subKeys = {};$invalidatedRoots.$parentKey = null;$invalidatedRoots.$parent = null;let $first = true;let $tainted = new WeakSet();$invalidatedMap.set($res, $invalidatedRoots);const untrack = ($targetKeySet, $targetKey) => {
      const $tracked = $trackedMap.get($targetKeySet);if (!$tracked || !$tracked[$targetKey]) {
        return;
      }$tracked[$targetKey].forEach(({ $sourceObj, $sourceKey, $target }) => {
        const $trackingSource = $trackingMap.get($sourceObj);$trackingSource[$sourceKey].delete($target);
      });delete $tracked[$targetKey];
    };const invalidate = ($targetKeySet, $targetKey) => {
      if ($targetKeySet.has($targetKey)) {
        return;
      }$targetKeySet.add($targetKey);untrack($targetKeySet, $targetKey);if ($targetKeySet.$parent) {
        invalidate($targetKeySet.$parent, $targetKeySet.$parentKey);
      }
    };function setOnObject($target, $key, $val, $invalidates) {
      let $changed = false;let $hard = false;if ($invalidates && !$first) {
        if (typeof $target[$key] === 'object' && $target[$key] && $target[$key] !== $val) {
          $hard = true;
        }if ($hard || $target[$key] !== $val || typeof $val === 'object' && $tainted.has($val) || !$target.hasOwnProperty($key) && $target[$key] === undefined) {
          $changed = true;triggerInvalidations($target, $key, $hard);
        }
      }$target[$key] = $val;return $changed;
    }function triggerInvalidations($sourceObj, $sourceKey, $hard) {
      $tainted.add($sourceObj);const $track = $trackingMap.get($sourceObj);if ($track && $track.hasOwnProperty($sourceKey)) {
        $track[$sourceKey].forEach(($soft, $target) => {
          if (!$soft || $hard) {
            invalidate($target[0], $target[1]);
          }
        });
      }if ($trackingWildcards.has($sourceObj)) {
        $trackingWildcards.get($sourceObj).forEach($targetInvalidatedKeys => {
          invalidate($targetInvalidatedKeys, $sourceKey);
        });
      }
    }$invalidatedRoots.add('first');function $firstBuild() {
      const newValue = $model[0];setOnObject($res, 'first', newValue, false);$invalidatedRoots.delete('first');return $res.first;
    }$invalidatedRoots.add('second');function $secondBuild() {
      const newValue = $model[1];setOnObject($res, 'second', newValue, false);$invalidatedRoots.delete('second');return $res.second;
    }let $inBatch = false;let $batchPending = [];function recalculate() {
      if ($inBatch) {
        return;
      }$invalidatedRoots.has('first') && $firstBuild();$invalidatedRoots.has('second') && $secondBuild();$tainted = new WeakSet();$first = false;$listeners.forEach(callback => callback());if ($batchPending.length) {
        $res.$endBatch();
      }
    }Object.assign($res, {}, { $startBatch: () => $inBatch = true, $endBatch: () => {
        $inBatch = false;if ($batchPending.length) {
          $batchPending.forEach(({ func, args }) => {
            func.apply($res, args);
          });$batchPending = [];recalculate();
        }
      }, $runInBatch: func => {
        $res.$startBatch();func();$res.$endBatch();
      }, $addListener: func => {
        $listeners.add(func);
      }, $removeListener: func => {
        $listeners.delete(func);
      } });recalculate();return $res;
  }return model;
}();
"
`;

exports[`macros template-literal: template-literal 1`] = `
"
const carmi = require('./macro')

const modelBuilder = carmi\`
  const {root} = require('../..')
  module.exports = {all: root.get('list'), first: root.get('list').get(0)}
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const modelBuilder = (function () {
  'use strict';
  function model($model, $funcLib, $batchingStrategy) {
    const $res = { $model };const $listeners = new Set();const $trackingMap = new WeakMap();const $trackedMap = new WeakMap();const $trackingWildcards = new WeakMap();const $invalidatedMap = new WeakMap();const $invalidatedRoots = new Set();$invalidatedRoots.$subKeys = {};$invalidatedRoots.$parentKey = null;$invalidatedRoots.$parent = null;let $first = true;let $tainted = new WeakSet();$invalidatedMap.set($res, $invalidatedRoots);const untrack = ($targetKeySet, $targetKey) => {
      const $tracked = $trackedMap.get($targetKeySet);if (!$tracked || !$tracked[$targetKey]) {
        return;
      }$tracked[$targetKey].forEach(({ $sourceObj, $sourceKey, $target }) => {
        const $trackingSource = $trackingMap.get($sourceObj);$trackingSource[$sourceKey].delete($target);
      });delete $tracked[$targetKey];
    };const invalidate = ($targetKeySet, $targetKey) => {
      if ($targetKeySet.has($targetKey)) {
        return;
      }$targetKeySet.add($targetKey);untrack($targetKeySet, $targetKey);if ($targetKeySet.$parent) {
        invalidate($targetKeySet.$parent, $targetKeySet.$parentKey);
      }
    };function setOnObject($target, $key, $val, $invalidates) {
      let $changed = false;let $hard = false;if ($invalidates && !$first) {
        if (typeof $target[$key] === 'object' && $target[$key] && $target[$key] !== $val) {
          $hard = true;
        }if ($hard || $target[$key] !== $val || typeof $val === 'object' && $tainted.has($val) || !$target.hasOwnProperty($key) && $target[$key] === undefined) {
          $changed = true;triggerInvalidations($target, $key, $hard);
        }
      }$target[$key] = $val;return $changed;
    }function triggerInvalidations($sourceObj, $sourceKey, $hard) {
      $tainted.add($sourceObj);const $track = $trackingMap.get($sourceObj);if ($track && $track.hasOwnProperty($sourceKey)) {
        $track[$sourceKey].forEach(($soft, $target) => {
          if (!$soft || $hard) {
            invalidate($target[0], $target[1]);
          }
        });
      }if ($trackingWildcards.has($sourceObj)) {
        $trackingWildcards.get($sourceObj).forEach($targetInvalidatedKeys => {
          invalidate($targetInvalidatedKeys, $sourceKey);
        });
      }
    }$invalidatedRoots.add('all');function $allBuild() {
      const newValue = $model[\\"list\\"];setOnObject($res, 'all', newValue, false);$invalidatedRoots.delete('all');return $res.all;
    }$invalidatedRoots.add('first');function $firstBuild() {
      const newValue = $res[\\"all\\"][0];setOnObject($res, 'first', newValue, false);$invalidatedRoots.delete('first');return $res.first;
    }let $inBatch = false;let $batchPending = [];function recalculate() {
      if ($inBatch) {
        return;
      }$invalidatedRoots.has('all') && $allBuild();$invalidatedRoots.has('first') && $firstBuild();$tainted = new WeakSet();$first = false;$listeners.forEach(callback => callback());if ($batchPending.length) {
        $res.$endBatch();
      }
    }Object.assign($res, {}, { $startBatch: () => $inBatch = true, $endBatch: () => {
        $inBatch = false;if ($batchPending.length) {
          $batchPending.forEach(({ func, args }) => {
            func.apply($res, args);
          });$batchPending = [];recalculate();
        }
      }, $runInBatch: func => {
        $res.$startBatch();func();$res.$endBatch();
      }, $addListener: func => {
        $listeners.add(func);
      }, $removeListener: func => {
        $listeners.delete(func);
      } });recalculate();return $res;
  }return model;
})();
"
`;
